src/databricks/labs/ucx/source_code/graph.py

```diff
@@ -261,7 +265,9 @@ def __repr__(self):
 
 class LibraryResolver(abc.ABC):
     @abc.abstractmethod
-    def register_library(self, path_lookup: PathLookup, library: Path) -> list[DependencyProblem]:
+    def register_library(
+        self, path_lookup: PathLookup, *libraries: str, installation_arguments: list[str] | None = None
```

Comment:
There are two changes in this API:
1. Multiple libraries can be installed at once. Motivation: with pip magic multiple installation can be installed at once. To remain as close as possible to the installation command defined as possible, thus minimizing the chance of mimicking the install incorrectly, we support installing multiple libraries at once using the installation arguments.
2. libraries are strings. Motivatioin: i) practical reason, when a library is defined as a relative path "./path/to/lib.whl" a conversion to `Path` loses the leading `./`, which makes that we can not replace the relatives anymore when resolving them with the path lookup. ii) here, libraries are strings, e.g. when `pip install pytest`, only later they are maybe a path when resolving them with the path lookup

Reply:
I think it's possible to make these changes while retaining compatibility:
```python
def register_library(self, path_lookup: PathLookup, library: Path, *more_libraries: Path, installation_arguments: list[str] | None = None) -> list[DependencyProblem]:
    ...
```
This type signature ensures that at least one argument is provided.

Can you elaborate a bit more on why there's a problem with the leading `./` being stripped? I'm a little confused because [anything without a leading `/` is considered relative](https://docs.python.org/3.10/library/pathlib.html#pathlib.PurePath.is_absolute), and if it's a problem why wasn't it a problem before?

Reply:
With current changes, I am replacing the resolved path in the installation arguments which uses an equality to [replace the argument](https://github.com/databrickslabs/ucx/blob/9be7b05a4b5d9016b824ceead254d5a9752b99e8/src/databricks/labs/ucx/source_code/python_libraries.py#L84)

```
args = [maybe_library.as_posix() if arg == library else arg for arg in args]
```

Before the library path was "just" resolved, the path lookup can handle both the relative paths with and without the ./

Reply:
I could split the `*libraries` into `library` and `libraries` to enforce at least defining one library. Though, it adds complexity to the code; the current code works; and allows users to unpack a possibly empty list of libraries. Though, it nudges users to define at least one library, which is can be a good thing. IMO only this change does not add much value

Reply:
this gets too complicated - just pass `list[str]` down to `pip` and let it handle all the work. also `shlex` them and download any `pathlib.Path`-like files